Marlin/i2cEncoder.cpp

```diff
@@ -0,0 +1,1173 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "Marlin.h"
+#include "i2cEncoder.h"
+#include "temperature.h"
+#include "stepper.h"
+
+#include <Wire.h>
+
+#if defined(I2C_ENCODERS_ENABLED)
+
+const char axis_codes[NUM_AXIS] = {'X', 'Y', 'Z', 'E'};
+
+void I2cEncoder::init(AxisEnum axis, byte address) {
+  set_axis(axis);
+  set_address(address);
+  initialised = true;
+  SERIAL_ECHO("Encoder on ");
+  SERIAL_ECHO(axis_codes[get_axis()]);
+  SERIAL_ECHO(" axis, address = ");
+  SERIAL_ECHOLN((int) address);
+  position = get_position();
+}
+
+void I2cEncoder::update() {
+
+  //check encoder is set up and active
+  if(initialised && homed && active) {
+
+    bool moduleDetected;
+
+    position = get_position();
+    //we don't want to stop things just because the encoder missed a message,
+    //so we only care about responses that indicate bad magnetic strength
+    bool signalGood = passes_test(false,moduleDetected);
+
+    //check encoder data is good
+    if(signalGood) {
+
+      //if data is historically good, proceed
+      if(trusted) {
+
+        //get latest position
+        lastPosition = position;
+        unsigned long positionTime = millis();
+
+        //only do error correction if setup and enabled
+        if(get_error_correct_method() != ECM_NONE && get_error_correct_enabled()) {
+
+          #if defined(ERROR_THRESHOLD_PROPORTIONAL_SPEED)
+            unsigned long distance = abs(position - lastPosition);
+            unsigned long deltaTime = positionTime - lastPositionTime;
+            unsigned long speed = distance / deltaTime;
+            float threshold = constrain((speed / 50),1,50) * get_error_correct_threshold();
+          #else
+            float threshold = get_error_correct_threshold();
+          #endif
+
+          //check error
+          #if ENABLED(ERROR_ROLLING_AVERAGE)
+            ((errArrayIndex >= ERROR_ARRAY_SIZE - 1) ? (errArrayIndex = 0) : errArrayIndex++);
+            errorArray[errArrayIndex] = get_axis_error_steps(false);
+            double  sum = 0;
+            double diffSum = 0;
+            for (int i=0;i<ERROR_ARRAY_SIZE;i++) {
+              sum += errorArray[i];
+              if (i > 0) diffSum += abs(errorArray[i-1]-errorArray[i]);
+            }
+            long error = (long)(sum/(ERROR_ARRAY_SIZE + 1)); //calculate average for error
+            
+          #else
+            long error = get_axis_error_steps(false);
+          #endif
+          
+          //SERIAL_ECHO("Axis err*r steps: ");
+          //SERIAL_ECHOLN(error);
+
+          #if defined(AXIS_ERROR_THRESHOLD_ABORT)
+            if(abs(error) > AXIS_ERROR_THRESHOLD_ABORT * planner.axis_steps_per_mm[encoderAxis]) {
+              //kill("Significant Error");
+              SERIAL_ECHO("Axis error greater than set threshold, aborting!");
```

Comment:
progmem... 

(okay, I'll stop mentioning these, though there are quite a few more later on in the code..)